apps/nifconv/nifconv.cpp

```diff
@@ -0,0 +1,155 @@
+#endif
+
+int main(int argc, char **argv) {
+    namespace bpo = boost::program_options;
+    bpo::options_description desc("Syntax: nifconv <options>\nAllowed options");
+
+    desc.add_options()
+        ("help", "print help message")
+        ("version", "print version information and quit")
+        ("fs-strict", bpo::value<bool>()->implicit_value(true)
+            ->default_value(false), "strict file system handling (no case folding)")
+        ("binary", bpo::value<bool>()->implicit_value(true)
+            ->default_value(false), "write OSG file in binary format (OSGB)")
+        ("data",
+         bpo::value<Files::PathContainer>()->default_value(Files::PathContainer(), "data")
+            ->multitoken()->composing(),
+         "set data directories (later directories have higher priority)")
+        ("data-local", bpo::value<std::string>()->default_value(""),
+         "set local data directory (highest priority)")
+        ("fallback-archive",
+         bpo::value<StringsVector>()->default_value(StringsVector(), "fallback-archive")
+         ->multitoken(), "set fallback BSA archives (later archives have higher priority)")
+        ("nif", bpo::value<StringsVector>(), "NIF file to convert")
+        ;
+
+    bpo::positional_options_description p;
+    p.add("nif", -1);
+
+    // Runtime options override settings from all configs
+    bpo::parsed_options valid_opts = bpo::command_line_parser(argc, argv)
+        .options(desc).positional(p).allow_unregistered().run();
+    bpo::variables_map variables;
+    bpo::store(valid_opts, variables);
+    bpo::notify(variables);
+
+    if (variables.count("help")) {
+        std::cout << desc << std::endl;
+        return false;
+    }
+
+    Files::ConfigurationManager cfgMgr;
+    cfgMgr.readConfiguration(variables, desc);
+
+    bool mFSStrict = variables["fs-strict"].as<bool>();
+    Files::PathContainer dataDirs(variables["data"].as<Files::PathContainer>());
+
+    std::string local(variables["data-local"].as<std::string>());
+    if (!local.empty()) {
+        dataDirs.push_back(Files::PathContainer::value_type(local));
+    }
+
+    cfgMgr.processPaths(dataDirs);
+
+    Files::Collections mFileCollections = Files::Collections (dataDirs, !mFSStrict);
+
+    std::vector<std::string> mArchives;
+    std::auto_ptr<Resource::ResourceSystem> mResourceSystem;
+
+    // fallback archives
+    StringsVector archives = variables["fallback-archive"].as<StringsVector>();
+    BOOST_FOREACH(const std::string a, archives) {
+        mArchives.push_back(a);
+    }
+
+    // engine.setResourceDir(variables["resources"].as<std::string>());
+        
+    std::auto_ptr<VFS::Manager> mVFS;
+    mVFS.reset(new VFS::Manager(mFSStrict));
+    VFS::registerArchives(mVFS.get(), mFileCollections, mArchives, true);
+    mResourceSystem.reset(new Resource::ResourceSystem(mVFS.get()));
+
+    Resource::SceneManager* sceneManager = mResourceSystem->getSceneManager();
+
+#ifdef ACTIVATE
```

Comment:
The consensus on the Internet seems to be that OSG trick for using global static object constructors to initialize an auto-magical plugin architecture is a BadIdea(tm).  For a long time I couldn't figure out how to get this to work at all because global objects are pulled in from static libraries _only_ if they're referenced in the main program on Linux.   Apparently the default behavior is different under Windows (first warning sign in my opinion).  To make matters worse, there was literally no way to reference the object from main in the standard OSG code, since the object had nothing but a constructor and destructor.   I ended up extending OSG with a null activate() method so that I could cause the object to be linked into the main program statically.  This code is no longer used since I figure out how to make the shared libraries properly, but I'm still uncertain if this is a "wise" architecture.